### ISSUE TITLE
Triumph of Ferocity log message.

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#531]
I have [Triumph of Ferocity](http://magiccards.info/query?q=!Triumph+of+Ferocity) in play which always generates the final message shown in the log.
- There is no link to ToF in the message.
- This is another case of a verbose log message when the ability was not actually activated - it would be better to just skip the message altogether I think, especially with no indication as to what card the message originated from.

![screenshot](https://cloud.githubusercontent.com/assets/2443519/11509107/18de4374-9854-11e5-8e2b-44554aa22334.jpg)

**Comment by ShawnieBoy** [comment]
There's no link as Triumph doesn't refer to itself, the same as if it was "At the beginning of your upkeep, you lose 1 life", the log message would only show "PN loses 1 life"

Triumph will always trigger, so needs to go onto the stack, it only checks whether you draw or not when it resolves. Unfortunately I don't think the DrawAction logs card draws, so it won't show if you draw or not either.

**Comment by lodici** [comment]
What I meant was wouldn't it be useful to know what card is generating that message? It had me flicking through the cards on the battlefield trying to find which one it was. It just seems a bit inconsistent - usually a log message has a reference to card (which now means an interactive link).

**Comment by melvinzhang** [comment]
MagicMessage doesn't keep track of the the card which created it, for example in the screen shot above "Destroy target artifact, enchantment, or land." It is not shown what is the source of this effect, only the selected choice (Sacred Foundry) is shown.

I think it is good to include it for clarity. The stack button includes the source at the top of the button as a title, hovering over the title pops up the card image. Perhaps something similar for the log?

**Comment by lodici** [comment]
> It is not shown what is the source of this effect, only the selected choice (Sacred Foundry) is shown.

Although you can infer it from the previous message. Not so much with the last message shown.

> The stack button includes the source at the top of the button as a title, hovering over the title pops up the card image. Perhaps something similar for the log?

If it is possible to prefix the message with "SN:" I think that would clarify what is generating the message and obviously also means an interactive link.

**Comment by ShawnieBoy** [comment]
I think using titles is the best way - if the source isn't referenced in an effect, it would become strange to try and fit it in.

SN makes PN draw a card? eww

**Comment by melvinzhang** [comment]
I don't think it needs to be part of the sentence. As @lodici suggested, it would be a prefix, eg
Triumph of Ferocity: Tester draws a card.

and it will only be for triggers

**Comment by melvinzhang** [comment]
The colon may be too close to the actual text. Perhaps something like "Triumph of Ferocity's ability triggers. Tester draws a card."

**Comment by lodici** [comment]
> Perhaps something like "Triumph of Ferocity's ability triggers. Tester draws a card."

That certainly reads better than the colon.

**Comment by lodici** [comment]
What would be displayed if the ability was not triggered?

**Comment by melvinzhang** [comment]
There will be no extra text, just what we see now in the stack/log.

**Comment by lodici** [comment]
So for Triumph of Ferocity you would have _"Triumph of Ferocity's ability triggers. Tester draws a card."_ or _"Tester draws a card if he or she controls the creature with the greatest power or tied for the greatest power"_?

**Comment by melvinzhang** [comment]
It would be

Triumph of Ferocity's ability triggers. Tester draws a card if he or she controls the creature with the greatest power or tied for the greatest power.

**Comment by melvinzhang** [comment]
My bad, I read your question "What would be displayed if the ability was not triggered?" as "What if it was not a triggered ability?". If the ability did not trigger there will be no stack/log item created.

**Comment by lodici** [comment]
> If the ability did not trigger there will be no stack/log item created.

Awesome :+1:

**Comment by melvinzhang** [comment]
Triumph of Ferocity ability triggers on your upkeep, it will always trigger when the condition is met, i.e. it is your upkeep.

**Comment by lodici** [comment]
Maybe I not using the correct terminology but I don't see the point of logging _"Tester draws a card if he or she controls the creature with the greatest power or tied for the greatest power"_ if the condition is not met. Just seems like noise to me.

**Comment by melvinzhang** [comment]
The condition is  "At the beginning of your upkeep". The effect is "draw a card if you control the creature with the greatest power or tied for the greatest power." The effect happens when the trigger resolves, you can cast a pump spell which will resolve before the trigger resolves and make your creature have the greatest power.

You might be thinking of the "intervening if" condition such as "Whenever SN deals damage, if you have seven or more cards in hand, you gain 7 life.". In the "intervening if" case, the ability will not trigger if the condition, eg "seven or more cards in hand", is not satisfied.

**Comment by lodici** [comment]
Ach! MTG overload! I not sure we are on the same page - all I am really questioning is why is the last log message displayed in the screenshot shown? What is the point of it? Most log messages record something being "done" - PN loses X life, SN is destroyed, Destroy target artifact..., etc. That all makes sense to me, but when I get to the last message I am confused as to whether something has been done or not, and if not why is it mentioning it? Obviously I am coming at this as a relative MTG novice but then I would say that is where a clear log really helps in explaining the mechanics.

**Comment by melvinzhang** [comment]
A simple way to think about it is when something on the stack resolves, it goes onto the log. When the event on the stack "Tester draws a card if he or she controls the creature with the greatest power or tied for the greatest power" resolves, it goes to the log.

Currently there is no mechanism to alter or modify the log message, it is directly taken from the event that went on the stack.

**Comment by melvinzhang** [comment]
Another way to put it is that most effects are written as "Do something", some others are templated as "Do something if condition". We do not treat them differently at this point, they go onto the log as they are written.

**Comment by melvinzhang** [comment]
I think it is important to indicate in the log that the ability triggered. The problem here is not indicating whether a card draw occurred or not. This is to differentiate from the case where the ability did not trigger at all, such as when the ability was lost.

**Comment by melvinzhang** [comment]
To keep this issue on track, I've split out the concern about the log being unclear if the action was performed into #536 

For this issue, the problem we are addressing is not knowing where an effect in the log comes from. This is particularly problematic for triggers. The suggestion was to add the prefix "SN's ability triggers. " in front of the original text.

Thinking this through, it is possible to do this automatically for triggers in card script by modifying MagicSourceEvent.getTriggerEvent. Unfortunately, for triggers in groovy code, the prefix needs to be added manually. 

It looks like trying to enhance the log only for triggers is difficult. It seems to be easier to enhance all log messages with the source.

**Comment by ShawnieBoy** [comment]
I have to agree with @melvinzhang that enhancing all the log messages would be better, there are also activated abilities which don't refer to the source (and don't tap, which is usually a visual clue)
[Cliffside Lookout](http://magiccards.info/bfz/en/20.html) for example.

Having a title-bar on the log messages would prevent having to edit the cards, and would help keep the Magic-ese style of the text.

As for the trigger message, it's similar in a way to #502 Where the trigger goes onto the stack, and then checks the condition when it resolves. As DrawCardAction doesn't send a message to the log, it's hard to tell if the effect did anything or not.

**Comment by lodici** [comment]
> Having a title-bar on the log messages would prevent having to edit the cards, and would help keep the Magic-ese style of the text.

It might make implementation easier but visually it would be quite a significant change to how the log looks. Using up a line in each item for a titlebar seems extravagant given the limited space available.

**Comment by ShawnieBoy** [comment]
>It might make implementation easier but visually it would be quite a significant change to how the log looks. Using up a line in each item for a titlebar seems extravagant given the limited space available.

But as a play aid, it's certainly needed - could also be used for combat damage, which currently appears with no other information - "PN loses X life"

I think a title-bar could look quite good

**Comment by ShawnieBoy** [comment]
From the log above
![titlebar](https://cloud.githubusercontent.com/assets/2363180/11532476/fe8ffcee-98fb-11e5-8891-5d2e3a07e56f.png)

**Comment by lodici** [comment]
Thanks for mock up - interesting. What about log items that do not have a source? For example, PN loses X life.

**Comment by ShawnieBoy** [comment]
I guess for combat damage, it would have the header 'Combat'

**Comment by ShawnieBoy** [comment]
I may be wrong, but everything will have a source, just that log messages don't always refer to the source (the example mock up I gave).
For the combat damage it could even be the creature that dealt the damage?

**Comment by melvinzhang** [comment]
Alternating black title bar and light gray text box may be a bit too striking. Seems that most of the time the source is not needed, except where it can be confusing. Another possibility is to pop up the card, when you hover over the log entry on the non-interactive segments, so it only appears when you need the info.

**Comment by ShawnieBoy** [comment]
>Alternating black title bar and light gray text box may be a bit too striking. Seems that most of the time the source is not needed, except where it can be confusing. Another possibility is to pop up the card, when you hover over the log entry on the non-interactive segments, so it only appears when you need the info.

Actually, isn't that what happens anyway?

**Comment by lodici** [comment]
> There's two other logs there with no source information - "Destroy target artifact, enchantment or land." and "Target creature gets +2/+2 and gains lifelink until end of turn."

These can be inferred from the previous log entries - Acidic Slime and Moment of Heroism. That is why the last entry in the screenshot stands out, there is no context.

> Alternating black title bar and light gray text box may be a bit too striking.

Good point - we probably need to suck it and see.

> Seems that most of the time the source is not needed, except where it can be confusing

Agree. The log should be concise. MTG is complicated enough!

> Another possibility is to pop up the card, when you hover over the log entry on the non-interactive segments, so it only appears when you need the info.

Sorry, not sure what you mean. 

If there was a titlebar I would be inclined to move the turn and phase into it, right justified, so there would be more room for the actual log message.

**Comment by lodici** [comment]
Did some hacking. Undecided. I can see some people might like it while others would prefer the cleaner non-titlebar look.

![testing](https://cloud.githubusercontent.com/assets/2443519/11534918/4008fb50-9909-11e5-9429-72a204fad11e.jpg)

**Comment by melvinzhang** [comment]
> Another possibility is to pop up the card, when you hover over the log entry on the non-interactive segments, so it only appears when you need the info.

Lemme redo that suggestion. I think we can add an icon to the log, same as the icon on the title bar of the stack button, i.e. StackViewerInfo.icon On the log, make the icon interactive, when you hover over it, it pops up the card image of the source (similar to hovering over the title bar on the stack).

I think it is better to put all the meta information to the side rather than to the top/bottom since the log is running down vertically it is hard to tell if the title is for the bottom or top log.

I think we can squeeze a bit more space out of the left side, by combining the heart and the life total to one icon and putting the source icon below it. In ascii art form, something like
```
PP L  <log text line 1>
PP S  <log text line 2>
```
where P is the player avatar, L is the life total number, S is the source icon.

**Comment by lodici** [comment]
I think there is an argument to display the source name if present so you can just read it without having to do anything. However, I have always wondered why there is a need to display life constantly in the log. Using your approach I would have the source icon first since it is likely to be present most frequently then below it display significant event icons as they happen such as loss/gain life etc.

**Comment by lodici** [comment]
See also #1112 for a similar example using Spirit Link.

**Comment by melvinzhang** [comment]
Source icon first is fine. We can put in the source name later when we find a good way to do so.
